Basestation_Software.Web/Core/Components/PingTool.razor

```diff
@@ -0,0 +1,339 @@
+            device.PingStatus = reply.Status.ToString();
+
+            // Set PingStatusHighlight
+            if (reply.Status == IPStatus.Success)
+            {
+                device.PingStatusHighlight = goodConnection_hightlight;
+                failedPings = 0;
+            }
+            else
+            {
+                device.PingStatusHighlight = poorConnection_hightlight;
+                failedPings += 1;
+            }
+
+            await InvokeAsync(StateHasChanged);
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine(ex);
+        }
+    }
+
+    // Handles automatic ICMP pinging of components
+    private bool AutoPingProcessRunning = false;
+    private async void AutoPing(object? sender, System.Timers.ElapsedEventArgs e)
+    {
+        if (!AutoPingProcessRunning) // Ensures that there is not already a process running
+        {
+            try
+            {
+                AutoPingProcessRunning = true;
+                autoPingTimer.Enabled = false;
+                foreach (Device device in devices)
+                {
+                    if (device.AutoPing == true)
+                    {
+                        await PingIpAddressAsync(device, device.IpAddress);
+
+                        if (failedPings >= 4) // Display alert modal if 4 or more consecutive pings fail
+                        {
+                            if (warnOfPoorConnection)
+                            {
+                                await InvokeAsync(poorConnection_Modal.Open);
+                            }
+                            failedPings = 0;
+                        }
+
+                        await Task.Delay(autoPingDelay);
+                        continue;
+                    }
+                }
+                AutoPingProcessRunning = false;
+                autoPingTimer.Enabled = true;
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
+            }
+        }
+    }
+
+    // Sets all components to be pinged automatically.
+    private void AutoPingAll(bool value)
+    {
+        foreach (Device device in devices)
+        {
+            device.AutoPing = value;
+        }
+    }
+
+    // Updates whether ping tool should warn user of a poor connection
+    private void UpdateWarnOfPoorConnection(bool value)
+    {
+        warnOfPoorConnection = value;
+    }
+
+    private void UpdateAutoPingDelay(string delay)
+    {
+        try
+        {
+            autoPingDelay = int.Parse(delay) * 1000;
+        }
+        catch
+        {
+            autoPingDelay = 5000;
+        }
+    }
+
+    private async void RovePingBoard(Device device, string board)
+    {
+        device.PingStatus = "Pinging...";
+        await InvokeAsync(StateHasChanged);
+
+        DateTime start = DateTime.Now;
+        RoveCommService.Ping(board);
+        var response = await RoveCommService.Listen<int>(RoveCommManifest.SystemPackets.PING_REPLY);
```

Comment:
Impressive that you figured out a way to do rovecomm pings without me actually providing a way to do it in the rovecomm API. Note that this will break if you 
1. click "Rove" for CoreBoard
2. click "Rove" for PMS
3. PMS responds first
4. now `RovePingBoard()` for both CoreBoard and PMS return successfully because a PING_REPLY arrived
5. CoreBoard itself never actually responds
This is fine though since I don't think we ever use the Rove ping lol.
 Maybe in the future, I'll give RoveComm a `RovePingResponse response = await RoveCommService.PingAsync(board)` API where `RovePingResponse` tells you what board gave the PING_REPLY